### PR TITLE
Update table styles for universal listings

### PR DESCRIPTION
**Pull request by gasman**
Update listing table styles for the new universal listing designs (https://github.com/wagtail/wagtail/discussions/10446).

* Make the 'more' button on page listings into a '...' icon (that doesn't disappear on hover)
* Move all buttons other than 'Edit' into the More menu
* Move buttons to the right of title, rather than below
* Adjust cell alignments to be top-aligned as standard
* Add folder icon for pages with children

I've also updated the pre-commit hooks - for some reason, the older version of stylelint fails on my machine with a JS syntax error.

Screenshot 2023-09-15 at 00 09 15

_Please check the following:_

-   [ ] Do the tests still pass?[^1]
-   [ ] Does the code comply with the style guide?
    -   [ ] Run `make lint` from the Wagtail root.
-   [ ] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [ ] **Please list the exact browser and operating system versions you tested**:
    -   [ ] **Please list which assistive technologies [^3] you tested**:
-   [ ] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

### CODE REVIEW

**Review by thibaudcolas — CHANGES_REQUESTED** [review]
All around this is looking very solid. There are a few aspects that need tweaking, particularly to support when page titles wrap to multiple lines.

File: client/scss/components/_listing.scss
Comment (gasman) [reply client/scss/components/_listing.scss]: Updated in 5df19e2cf2613e54271fdc3b33c129e640783660.

File: wagtail/admin/tests/test_buttons_hooks.py
Comment (gasman) [reply wagtail/admin/tests/test_buttons_hooks.py:187]: Good shout... removed in 95290388a520ebe876a9823c290f4c497af401ed.

File: wagtail/admin/templates/wagtailadmin/pages/listing/_page_title_explore.html
Comment (gasman) [reply wagtail/admin/templates/wagtailadmin/pages/listing/_page_title_explore.html]: Have removed the utility class in 4a7e46f68892c830c90be6225e0f3d89e2dbaf80 - that was just a bit of casual copy-and-paste from the site icon...

The alignment was my attempt at getting the bottom of the folder to line up with the baseline of the text:
Screenshot 2023-09-18 at 16 59 41

Anything else looked off to me, but happy to defer to others on the design details here. (As I recall, the alignment on the globe icon for sites was way off before I made this fix, so I figure anything is an improvement :-) )

> With this setup, I see the site’s "Root" level gets a folder icons, while homepages do not. Is that what we’re after?

If by 'site’s "Root" level' you mean immediate children of the homepage - yes, I think it is. From an editor's perspective, I think "this is a site", "this is a folder" and "this is an ordinary page" are seen as mutually exclusive cases. While technically the site root could also be considered a folder on the basis that it has children, that feels more like programmer reasoning than something useful to editors getting their bearings.

File: wagtail/admin/wagtail_hooks.py
Comment (gasman) [reply wagtail/admin/wagtail_hooks.py:243]: Done in d7a148c9d380627b8ef866135d944a80dd3d62b0 - although this may only be a short-lived fix, as the designs call for "Add child page" to be pulled out into its own action in the header.

File: client/scss/components/_listing.scss
Comment (gasman) [reply client/scss/components/_listing.scss:310]: I guess this could go either way really... I took the plunge of setting top alignment globally (in d758bd41d9be8b1dc8dda355673cb0f72bf99e7e) so that tables with one overly tall column (e.g. the People snippet index on bakerydemo, which has a thumbnail image in it) don't end up shifting the other cells' contents weirdly - and also because we were setting it on an ad-hoc basis on a load of tables, to the point where it seemed we were fighting against a bad default. I guess this was me following that design decision through to its conclusion - but maybe this is a sensible exception to the rule, since it's treating the table cell as more of a button than a content element. (On the other hand, the new design - not implemented yet as it requires being able to calculate the descendant count - does bring it back into line with the 'link' style elsewhere in the table. Hmm...)

File: wagtail/admin/wagtail_hooks.py
Comment (gasman) [reply wagtail/admin/wagtail_hooks.py:226]: I'm fine with it... I can see there's an argument for standardising on the '...' icon for 'More' dropdowns, but I don't think that's a massively high priority.

File: wagtail/admin/views/pages/usage.py
Comment (gasman) [reply wagtail/admin/views/pages/usage.py:35]: Added a `margin: 0` on UL elements in 8619385f6620291a8a49867a32d484104c758df2 to address this. Hopefully there won't be any knock-on effects...

File: client/scss/components/_listing.scss
Comment (gasman) [reply client/scss/components/_listing.scss]: Probably worth highlighting that 5df19e2cf2613e54271fdc3b33c129e640783660 makes action buttons always visible for all listings, so it adds a lot of visual heaviness to tables elsewhere (e.g. snippet index and the "pages for moderation" dashboard panel). Do we want to hold off on this until those have received design attention too? Or is it justified to go ahead with it anyway, given that it's an accessibility issue?

File: client/scss/components/_listing.scss
Comment (gasman) [reply client/scss/components/_listing.scss:237]: Maybe this is just my lack of CSS-fu, but the only way I know of doing that is to move the `<ul class="actions">` before the `<div class="title-wrapper">` so that the title wraps around it - and I suspect that's an accessibility no-no (because we want screen-readers to read the title before the actions).

(Also, it's worth considering that wrapping onto a second line might happen because there are an excessive number of action buttons, and in that case it would make more sense to shift the buttons down. However, now that we've moved Edit into the More menu, and custom hooks / third-party packages are the only route by which we'll get any buttons here at all, this is probably enough of an edge case that we can disregard it...)

File: client/scss/components/_listing.scss
Comment (gasman) [reply client/scss/components/_listing.scss]: Hopefully fixed in 287e25d68d814ff2e6631da164042aff928b2432 - the aim was to match the height of the button (2em), and I'm not sure how the `rem` crept in... changing it to `em` still keeps the button centre-aligned, and multi-line titles are looking more sensible.
Screenshot 2023-09-18 at 18 59 51

File: client/scss/components/_listing.scss
Comment (thibaudcolas) [reply client/scss/components/_listing.scss:311]: 👍 what we have right now seems reasonable to me for this iteration of the designs. We should probably ask Ben to design a wrapping version and see where he takes it.

**Review by thibaudcolas — CHANGES_REQUESTED** [review]
@gasman the new listing styles are looking good! My only remaining concern is with linter versions in `.pre-commit-config.yaml` – could you move that to a separate PR so we can move on with those listing styles?

In particular on linters,

- We want to make sure the versions keep matching between pre-commit, `package.json`, `setup.py`
- For stylelint specifically, we need to update the version in two places in the hook definition, and would also likely want to update our `@wagtail/stylelint-config-wagtail` at the same time

**Review by thibaudcolas — CHANGES_REQUESTED** [review]
@gasman the new listing styles are looking good! My only remaining concern is with linter versions in `.pre-commit-config.yaml` – could you move that to a separate PR so we can move on with those listing styles?

In particular on linters,

- We want to make sure the versions keep matching between pre-commit, `package.json`, `setup.py`
- For stylelint specifically, we need to update the version in two places in the hook definition, and would also likely want to update our `@wagtail/stylelint-config-wagtail` at the same time

**Comment by gasman** [comment]
@thibaudcolas Have removed the linter update commit (now opened as #10928) and rebased.

**Review by thibaudcolas — APPROVED** [review]
Yay!

**Pull request merged**
